rbd/admin/admin.go

```diff
@@ -0,0 +1,63 @@
+// +build !nautilus
+
+package admin
+
+import (
+	"fmt"
+
+	ccom "github.com/ceph/go-ceph/common/commands"
+)
+
+// RBDAdmin is used to administrate rbd volumes and pools.
+type RBDAdmin struct {
+	conn ccom.RadosCommander
+}
+
+// NewFromConn creates an new management object from a preexisting
+// rados connection. The existing connection can be rados.Conn or any
+// type implementing the RadosCommander interface.
+func NewFromConn(conn ccom.RadosCommander) *RBDAdmin {
+	return &RBDAdmin{conn}
+}
+
+// LevelSpecifier provides an interface that can be used where ceph requires a
+// levelspec.
+type LevelSpecifier interface {
+	levelSpec() string
+}
+
+// RawLevelSpec can be used to pass a level spec directly to ceph without
+// constructing it from component values. This should only be used if LevelSpec
+// can not represent the levelspec value you want to pass to ceph.
+type RawLevelSpec string
+
+func (r RawLevelSpec) levelSpec() string {
+	return string(r)
+}
+
+// LevelSpec is used to construct a levelspec given a pool and
+// optional Namespace and Image names.
+type LevelSpec struct {
+	Pool      string
+	Namespace string
+	Image     string
+}
+
+func (ls LevelSpec) levelSpec() string {
+	return ls.String()
+}
+
+// String returns a string representation of the LevelSpec.
+func (ls LevelSpec) String() string {
+	var s string
+	if ls.Image != "" && ls.Namespace != "" {
+		s = fmt.Sprintf("%s/%s/%s", ls.Pool, ls.Namespace, ls.Image)
+	} else if ls.Image != "" {
+		s = fmt.Sprintf("%s/%s", ls.Pool, ls.Image)
+	} else if ls.Namespace != "" {
+		s = fmt.Sprintf("%s/%s/", ls.Pool, ls.Namespace)
+	} else {
+		s = fmt.Sprintf("%s/", ls.Pool)
+	}
+	return s
+}
```

Comment:
So, my opinion is still, that this is a too bloated API for such a simple thing. We export a lot of stuff (an interface, two types and a method) that increases the maintenance surface without a real need. I can't prove it really, but I think this would rather bite us than help us later. I wouldn't veto, but prepare for a "I told you so" later. ;-)

To be more constructive, here my counter proposal:

```go
// LevelSpecifier provides an interface that can be used where ceph requires a
// levelspec.
type LevelSpec struct {
	s string
}

// NewLevelSpec is used to construct a LevelSpec given a pool and
// optional Namespace and Image names.
func NewLevelSpec(pool, namespace, image string) LevelSpec {
	var s string
	if image != "" && namespace != "" {
		s = fmt.Sprintf("%s/%s/%s", pool, namespace, image)
	} else if image != "" {
		s = fmt.Sprintf("%s/%s", pool, image)
	} else if namespace != "" {
		s = fmt.Sprintf("%s/%s/", pool, namespace)
	} else {
		s = fmt.Sprintf("%s/", pool)
	}
	return LevelSpec{s}
}

// NewLevelSpecRaw can be used to pass a level spec directly to ceph without
// constructing it from component values. This should only be used if NewLevelSpec
// can not create the levelspec value you want to pass to ceph.
func NewLevelSpecRaw(spec string) LevelSpec {
	return LevelSpec{spec}
}
```

Reply:
That suggestion seems reasonable to me, although I don't see it a huge difference. :-) 
I have no problems incorporating this version. It has the advantage of keeping the fields entirely private, so if we did have to do something more complex in the future it's just a matter of changing the internal implementation. So I can't complain there. :-)

Reply:
done!